src/auth/src/credentials/external_account.rs

```diff
@@ -0,0 +1,258 @@
+            metadata: None,
+        };
+        Ok(token)
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct ExternalAccountCredentials<T>
+where
+    T: CachedTokenProvider,
+{
+    token_provider: T,
+    quota_project_id: Option<String>,
+}
+
+/// A builder for constructing external account [Credentials] instances.
+///
+/// # Example
+/// ```
+/// # use google_cloud_auth::credentials::external_account::{Builder};
+/// # tokio_test::block_on(async {
+/// let config = serde_json::json!({
+///     "type": "external_account",
+///     "audience": "//iam.googleapis.com/projects/<PROJECT_ID>/locations/global/workloadIdentityPools/<WORKLOAD_IDENTITY_POOL>/providers/<WORKLOAD_IDENTITY_PROVIDER_ID>",
+///     "subject_token_type": "urn:ietf:params:oauth:token-type:jwt",
+///     "token_url": "https://sts.googleapis.com/v1beta/token",
+///     "credential_source": {
+///         "url": "http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https://iam.googleapis.com/projects/<PROJECT_ID>/locations/global/workloadIdentityPools/<WORKLOAD_IDENTITY_POOL>/providers/<WORKLOAD_IDENTITY_PROVIDER_ID>",
+///         "headers": {
+///           "Metadata": "True"
+///         },
+///         "format": {
+///           "type": "json",
+///           "subject_token_field_name": "access_token"
+///         }
+///     }
+/// });
+/// let credentials = Builder::new(config)
+///     .build();
+/// })
+/// ```
+pub struct Builder {
+    external_account_config: Value,
+    quota_project_id: Option<String>,
+}
+
+impl Builder {
+    /// Creates a new builder using [external_account_config] JSON value.    
+    ///
+    /// [external_account_config]: https://cloud.google.com/iam/docs/workload-download-cred-and-grant-access#download-configuration
+    pub fn new(external_account_config: Value) -> Self {
+        Self {
+            external_account_config,
+            quota_project_id: None,
+        }
+    }
+
+    /// Sets the [quota project] for this credentials.
+    ///
+    /// In some services, you can use a service account in
+    /// one project for authentication and authorization, and charge
+    /// the usage to a different project. This requires that the
+    /// service account has `serviceusage.services.use` permissions on the quota project.
+    ///
+    /// [quota project]: https://cloud.google.com/docs/quotas/quota-project
+    pub fn with_quota_project_id<S: Into<String>>(mut self, quota_project_id: S) -> Self {
+        self.quota_project_id = Some(quota_project_id.into());
+        self
+    }
+
+    fn build_token_provider(self) -> Result<ExternalAccountTokenProvider> {
+        let external_account_config: ExternalAccountConfig =
+            serde_json::from_value(self.external_account_config).map_err(errors::non_retryable)?;
+        let token_provider = ExternalAccountTokenProvider::new(external_account_config)
+            .map_err(errors::non_retryable)?;
+        Ok(token_provider)
+    }
```

Comment:
Consider moving private functions to the bottom of the type. Easier to grok what the public methods are if they are all together (we have been bad at doing this elsewhere).